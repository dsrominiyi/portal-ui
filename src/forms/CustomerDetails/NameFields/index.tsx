import React, { FunctionComponent, Fragment } from 'react';

import QuestionBox from '../../../components/QuestionBox';
import TextField from '../../../components/TextField';

import { predefinedRules, ValidationError } from '../../../helpers/formValidator';

import formStyles from '../../styles.scss';

interface FieldValues {
  forename: string;
  middleName: string;
  surname: string;
}

interface FieldErrors {
  forename: ValidationError;
  middleName: ValidationError;
  surname: ValidationError;
}

export interface Props extends FieldValues {
  fieldErrors?: FieldErrors;
  onChange(values: FieldValues): void;
}

export const validationRules = {
  forename: { ...predefinedRules.individualName, fieldLabel: 'first name', required: true },
  middleName: { ...predefinedRules.individualName, fieldLabel: 'middle name' },
  surname: { ...predefinedRules.individualName, fieldLabel: 'surname', required: true }
};

export const NameFields: FunctionComponent<Props> = ({
  forename,
  middleName,
  surname,
  fieldErrors = {} as FieldErrors,
  onChange
}: Props) => (
  <Fragment>
    <QuestionBox
      hasError={fieldErrors.forename?.showError}
      errorText={fieldErrors.forename?.message}
      inputFields={[
        <TextField
          key="0"
          label="First name"
          value={forename}
          additionalClasses={[formStyles.formTextField]}
          isInvalid={fieldErrors.forename?.showError}
          onChange={event => onChange({ forename: event.target.value, middleName, surname })}
          attributes={{ 'test-id': 'first-name' }}
        />
      ]}
    />
    <QuestionBox
      hasError={fieldErrors.middleName?.showError}
      errorText={fieldErrors.middleName?.message}
      inputFields={[
        <TextField
          key="0"
          label="Middle name (optional)"
          value={middleName}
          additionalClasses={[formStyles.formTextField]}
          isInvalid={fieldErrors.middleName?.showError}
          onChange={event => onChange({ forename, middleName: event.target.value, surname })}
          attributes={{ 'test-id': 'middle-name' }}
        />
      ]}
    />
    <QuestionBox
      hasError={fieldErrors.surname?.showError}
      errorText={fieldErrors.surname?.message}
      inputFields={[
        <TextField
          key="0"
          label="Surname"
          value={surname}
          additionalClasses={[formStyles.formTextField]}
          isInvalid={fieldErrors.surname?.showError}
          onChange={event => onChange({ forename, middleName, surname: event.target.value })}
          attributes={{ 'test-id': 'surname' }}
        />
      ]}
    />
  </Fragment>
);

NameFields.displayName = 'NameFields';
export default NameFields;
