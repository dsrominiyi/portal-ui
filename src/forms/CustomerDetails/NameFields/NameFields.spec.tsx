import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import NameFields, { Props } from '.';

let fields: ShallowWrapper<Props>;
let forenameField: any;
let middleNameField: any;
let surnameField: any;

const forename = 'John';
const middleName = 'Dave';
const surname = 'Smith';

const onChange = jest.fn();
const fieldErrors = {
  forename: { showError: true, message: 'error!' },
  middleName: { showError: true, message: 'error!' },
  surname: { showError: true, message: 'error!' }
};

const update = () => {
  [forenameField, middleNameField, surnameField] = (fields.props() as any).children;
};

describe('NameFields', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    fields = shallow(<NameFields forename="" middleName="" surname="" onChange={onChange} />);
    update();
  });

  test('NameFields matches snapshot', () => {
    expect(fields).toMatchSnapshot();
  });

  test('NameFields with field errors matches snapshot', () => {
    fields.setProps({ fieldErrors });

    expect(fields).toMatchSnapshot();
  });

  test('updating the forename field calls onChange with the updated value', () => {
    forenameField.props.inputFields[0].props.onChange({ target: { value: forename } });

    expect(onChange).toHaveBeenCalledWith({ forename, middleName: '', surname: '' });
  });

  test('updating the middleName field calls onChange with the updated value', () => {
    fields.setProps({ forename });
    update();
    middleNameField.props.inputFields[0].props.onChange({ target: { value: middleName } });

    expect(onChange).toHaveBeenCalledWith({ forename, middleName, surname: '' });
  });

  test('updating the surname field calls onChange with the updated value', () => {
    fields.setProps({ forename, middleName });
    update();
    surnameField.props.inputFields[0].props.onChange({ target: { value: surname } });

    expect(onChange).toHaveBeenCalledWith({ forename, middleName, surname });
  });
});
