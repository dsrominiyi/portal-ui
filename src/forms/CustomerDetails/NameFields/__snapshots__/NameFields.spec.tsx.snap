// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameFields NameFields matches snapshot 1`] = `
<Fragment>
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "first-name",
            }
          }
          label="First name"
          onChange={[Function]}
          value=""
        />,
      ]
    }
  />
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "middle-name",
            }
          }
          label="Middle name (optional)"
          onChange={[Function]}
          value=""
        />,
      ]
    }
  />
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "surname",
            }
          }
          label="Surname"
          onChange={[Function]}
          value=""
        />,
      ]
    }
  />
</Fragment>
`;

exports[`NameFields NameFields with field errors matches snapshot 1`] = `
<Fragment>
  <QuestionBox
    errorText="error!"
    hasError={true}
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "first-name",
            }
          }
          isInvalid={true}
          label="First name"
          onChange={[Function]}
          value=""
        />,
      ]
    }
  />
  <QuestionBox
    errorText="error!"
    hasError={true}
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "middle-name",
            }
          }
          isInvalid={true}
          label="Middle name (optional)"
          onChange={[Function]}
          value=""
        />,
      ]
    }
  />
  <QuestionBox
    errorText="error!"
    hasError={true}
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "surname",
            }
          }
          isInvalid={true}
          label="Surname"
          onChange={[Function]}
          value=""
        />,
      ]
    }
  />
</Fragment>
`;
